src/disclosure-manager/DisclosureManager.jsx

```diff
@@ -0,0 +1,568 @@
+   */
+  resolveDismissChecksInSequence(keys) {
+    return new Promise((resolve, reject) => {
+      if (!keys.length) {
+        resolve();
+        return;
+      }
+
+      const key = keys.pop();
+
+      this.generatePopFunction(key)()
+        .then(() => {
+          this.resolveDismissChecksInSequence(keys).then(resolve).catch(reject);
+        })
+        .catch(() => {
+          reject();
+        });
+    });
+  }
+
+  /**
+   * Closes the disclosure after calling all checks and resolving all deferred promises.
+   * @return A Promise, resolved if the close was performed or rejected if it was not.
+   */
+  safelyCloseDisclosure() {
+    const disclosureKeys = Object.assign([], this.state.disclosureComponentKeys);
+
+    /**
+     * Before closing the disclosure, the dismiss checks for components in the stack are
+     * executed in stack order. Components will be dismissed in order up until a rejection occurs, at which point
+     * the blocking component will be presented.
+     */
+    return this.resolveDismissChecksInSequence(disclosureKeys).then(() => {
+      this.dismissChecks = {};
+      this.closeDisclosure();
+    });
+  }
+
+  /**
+   * Creates an instance of a pop function for the component represented by the given key.
+   */
+  generatePopFunction(key) {
+    return () => {
+      const { disclosureComponentKeys } = this.state;
+
+      if (disclosureComponentKeys[disclosureComponentKeys.length - 1] !== key) {
+        /**
+         * If the top component key in the disclosure stack does not match
+         * the key used to generate this function, or the key is undefined, then the pop action is rejected.
+         */
+        return Promise.reject();
+      }
+
+      let promiseRoot = Promise.resolve();
+      const dismissCheck = this.dismissChecks[key];
+
+      if (dismissCheck) {
+        promiseRoot = dismissCheck();
+      }
+
+      const callbackfunction = () => {
+        this.dismissChecks[key] = undefined;
+        this.resolveDismissPromise(key);
+      };
+
+      return promiseRoot
+        .then(() => {
+          this.popDisclosure(callbackfunction);
+        });
+    };
+  }
+
+  generateDisclosureComponentMappingForRender() {
+    const { navigationPromptResolutionOptions } = this.props;
+
+    const {
+      disclosureComponentKeys,
+      disclosureComponentData,
+      disclosureComponentDelegates,
+    } = this.state;
+
+    return disclosureComponentKeys.reduce((accumulator, key, index) => {
+      const componentData = disclosureComponentData[key];
+
+      accumulator[key] = {
+        component: (
+          <DisclosureManagerHeaderAdapterContext.Provider value={componentData.headerAdapterContextValue} key={key}>
+            <DisclosureManagerContext.Provider value={disclosureComponentDelegates[index]}>
+              <DisclosureContainer navigationPromptResolutionOptions={navigationPromptResolutionOptions}>
```

Comment:
Would it be worth adding new disclosure manager props to the existing disclosure manger/slide-panel-manager & modal manger instead of reimplementing them?

```diff
// in terrra-disclosure-manager
....
- generateDisclosureComponentMappingForRender() {
+ generateDisclosureComponentMappingForRender(DisclosureContainer) {
    const { navigationPromptResolutionOptions } = this.props;

    const {
      disclosureComponentKeys,
      disclosureComponentData,
      disclosureComponentDelegates,
    } = this.state;

    return disclosureComponentKeys.reduce((accumulator, key, index) => {
      const componentData = disclosureComponentData[key];

      accumulator[key] = {
        component: (
          <DisclosureManagerHeaderAdapterContext.Provider value={componentData.headerAdapterContextValue} key={key}>
            <DisclosureManagerContext.Provider value={disclosureComponentDelegates[index]}>
-              {componentData.component}
+             <DisclosureContainer component={componentData.component} >
            </DisclosureManagerContext.Provider>
          </DisclosureManagerHeaderAdapterContext.Provider>
        ),
        headerAdapterData: componentData.headerAdapterData,
      };

      return accumulator;
    }, {});
  }
...
  render() {
-    const { render, children } = this.props;
+    const { render, disclosureContainer, children } = this.props;

....
-    const disclosureComponentMappingForRender = this.generateDisclosureComponentMappingForRender();
+    const disclosureComponentMappingForRender = this.generateDisclosureComponentMappingForRender(disclosureContainer);
```